### PR TITLE
IEx: Sort Types in `t` helper

### PR DESCRIPTION
**Pull request by maennchen**
Fix for #9129

### CODE REVIEW

File: lib/iex/lib/iex/introspection.ex [hunk 1]
```diff
@@ -644,7 +644,16 @@ defmodule IEx.Introspection do
         types_not_found(inspect(module))
 
       {:ok, types} ->
-        Enum.each(types, &(&1 |> format_type() |> IO.puts()))
+        types
+        |> Enum.filter(&match?({kind, _} when kind in [:type, :opaque], &1))
```
Comment (maennchen) [inline lib/iex/lib/iex/introspection.ex]: @josevalim I also filtered out `typep` here, which was not done before.

Comment (wojtekmach) [reply lib/iex/lib/iex/introspection.ex]: According to #2612 we want to list typeps so Id be careful about making this particular change.

Comment (fertapric) [reply lib/iex/lib/iex/introspection.ex]: I think we should add a test to prevent `typep` to be filtered out in future PRs

Comment (maennchen) [reply lib/iex/lib/iex/introspection.ex]: @wojtekmach That's why I'm specifically pointing it out. Should I leave it in or should I remove the line?

Before this PR we've shown `typep` for `t(module)`, but not for `t({module, type})` / `t({module, type, arity})`. This didn't make sense to me, therefore I changed the behaviour of the first to reflect the others.

Comment (maennchen) [reply lib/iex/lib/iex/introspection.ex]: @fertapric If the behavior from before is the desired one, I'll remove the line and can also add a test to make sure it is not filtered out in the future.

**Comment by josevalim** [comment]
Good call @wojtekmach.
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by josevalim** [comment]
Let’s remove the filtering from this PR for now and handle it in another PR.
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

File: lib/iex/lib/iex/introspection.ex (hunk 1 above)
Comment (maennchen) [reply lib/iex/lib/iex/introspection.ex]: @fertapric & @wojtekmach Adressed in 841f5ab

Comment (wojtekmach) [reply lib/iex/lib/iex/introspection.ex]: @maennchen @fertapric thanks for adding the test <3

File: lib/iex/lib/iex/introspection.ex [hunk 2]
```diff
@@ -644,7 +644,15 @@ defmodule IEx.Introspection do
         types_not_found(inspect(module))
 
       {:ok, types} ->
-        Enum.each(types, &(&1 |> format_type() |> IO.puts()))
+        types
+        |> Enum.sort_by(fn {_, {name, _, args}} -> {name, length(args)} end)
+        |> case do
+          [] ->
```
Comment (fertapric) [inline lib/iex/lib/iex/introspection.ex]: I think this clause can never be executed :)

Comment (maennchen) [reply lib/iex/lib/iex/introspection.ex]: Without the `typep` filter, you're correct. I'll remove the case.

Comment (maennchen) [reply lib/iex/lib/iex/introspection.ex]: Fixed in 9830d3d8089f6594a821a5fe39ccf2be0247d725

File: lib/iex/lib/iex/introspection.ex [hunk 3]
```diff
@@ -677,16 +683,24 @@ defmodule IEx.Introspection do
...
-        if printed == [] do
-          types_not_found_or_private("#{inspect(module)}.#{type}")
+        types
+        |> Enum.filter(
+          &match?(
+            {kind, {^type, _, args}} when kind in [:type, :opaque] and length(args) == arity,
+            &1
+          )
+        )
+        |> Enum.sort_by(fn {_, {name, _, args}} -> {name, length(args)} end)
```
Comment (fertapric) [inline lib/iex/lib/iex/introspection.ex]: I think you cannot have more than one result here for `{module, type, arity}`, @josevalim?

**Comment by josevalim** [comment]
Yeah, you can’t have more than one result for a given module, function,
arity.
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

File: lib/iex/lib/iex/introspection.ex (hunk 3 above)
Comment (maennchen) [reply lib/iex/lib/iex/introspection.ex]: @fertapric Adressed in 4250b33c872169fc5b4d57264b22c83b1701a3ab

Comment (fertapric) [reply lib/iex/lib/iex/introspection.ex]: The `Enum.sort_by` can also be removed

Comment (maennchen) [reply lib/iex/lib/iex/introspection.ex]: @fertapric Adressed in 1a6c82f5685b771a9766aafca9215bb39d37c6df

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**